### PR TITLE
Fix for #5061 adding toggleFilterControl () method

### PR DESCRIPTION
**Pull request by marceloverdijk**
This PR is for issue #5061 adding a `toggleFilterControl ()` method.

### CODE REVIEW

**Comment by UtechtDustin** [comment]
@marceloverdijk 
Please provide us an example using our [editor](live.bootstrap-table.com/).

**Comment by marceloverdijk** [comment]
Hi @UtechtDustin can the example use the code from the PR? Or do you expect something else?

**Comment by UtechtDustin** [comment]
@marceloverdijk yes, its possible.
Go on the Libraries page, then select `From GitHub src` on the option `Bootstrap Table source`.
Now you can set your Branch (with the repository) `marceloverdijk:toggle_filter_control` on the `GitHub src branch`.
After that you can create your example, save it and post the url of that example here.

**Comment by marceloverdijk** [comment]
Great! I wasn’t aware. Will try to do it this evening.

**Comment by marceloverdijk** [comment]
@UtechtDustin here is the example: https://live.bootstrap-table.com/code/marceloverdijk/3261

**Comment by UtechtDustin** [comment]
There is a issue while filtering and then toggle the filter control.
Peek 2020-06-07 22-47

**Comment by marceloverdijk** [comment]
Interesting, it works in my local project but indeed I can reproduce it in the example.
The setup of my table in my local project is a bit different. Let me check...

**Comment by marceloverdijk** [comment]
Oke, this one works https://live.bootstrap-table.com/code/marceloverdijk/3262 but has the data define using `<tbody>` which I also have in my local project.

The [first (failing) example](https://live.bootstrap-table.com/code/marceloverdijk/3261) uses `data-url="json/data1.json"`.
So the issue is related to that and at least explains why it was working in my local project.
I'm new to bootstrap-table and did not use the `data-url` yet, do you maybe have an idea what the issue could be?

**Comment by UtechtDustin** [comment]
No sorry, i would also have to debug that.

**Comment by UtechtDustin** [comment]
Could be solution but we also have to "reset" the filter-control functions.
Because if you would filter for something, and then toggle the filter-controle (hide the controls) the table would be breaken (The filtered data would be displayed).

**Comment by djhvscf** [comment]
> Could be solution but we also have to "reset" the filter-control functions.
> Because if you would filter for something, and then toggle the filter-controle (hide the controls) the table would be breaken (The filtered data would be displayed).

Yes, I think we can "reset" the filter controls and then hide/show it

**Comment by marceloverdijk** [comment]
Yes I certainly agree this is a better approach then refresh options.

But there is more than only clearing the filters and then hide/show them.
Some users want the filter controls to be disabled by default. 
With refresh options this is possible by setting `data-filter-control = false` (or off course not setting the property at all). The refresh option will draw the filter controls, but if the filter controls are not drawn `data-filter-control = false` we cannot show them as they don't exists.

Should we introduce a new `data-filter-control-hidden` property to be able to show/hide the filter controls on init?

`data-filter-control = true` ==> show the filter control
`data-filter-control = true && data-filter-control = true` ==> hide filter control but create elements

**Comment by djhvscf** [comment]
> Yes I certainly agree this is a better approach then refresh options.
> 
> But there is more than only clearing the filters and then hide/show them.
> Some users want the filter controls to be disabled by default.
> With refresh options this is possible by setting `data-filter-control = false` (or off course not setting the property at all). The refresh option will draw the filter controls, but if the filter controls are not drawn `data-filter-control = false` we cannot show them as they don't exists.
> 
> Should we introduce a new `data-filter-control-hidden` property to be able to show/hide the filter controls on init?
> 
> `data-filter-control = true` ==> create and show the filter controls
> `data-filter-control = true && data-filter-control = true` ==> create filter controls but hide them

Well in that case for me the data-filter-control-hidden is an overkill since you can simulate that in column configurations (by not setting the filter control)

**Comment by UtechtDustin** [comment]
> Well in that case for me the data-filter-control-hidden is an overkill since you can simulate that in column configurations (by not setting the filter control)

As he already explaint the filter-control extension will not create Controls, events etc.
Which means we cant toggle that (without recreating the whole table).

So we have 2 ways to go:
- Use a new option e.g. `data-filter-control-hidden` to init the filter-control extension but dont show it.
- Recreate the whole table and just toggle on/off the filter-control extension.

**Comment by djhvscf** [comment]
> > Well in that case for me the data-filter-control-hidden is an overkill since you can simulate that in column configurations (by not setting the filter control)
> 
> As he already explaint the filter-control extension will not create Controls, events etc.
> Which means we cant toggle that (without recreating the whole table).
> 
> So we have 2 ways to go:
> 
> * Use a new option e.g. `data-filter-control-hidden` to init the filter-control extension but dont show it.
> * Recreate the whole table and just toggle on/off the filter-control extension.

In that case I would prefer use the new option 'data-filter-control-hidden'.

**Comment by marceloverdijk** [comment]
I refactored `toggleFilterControl` to not call the refresh but show/hide the element.
Also the new `data-filter-control-hidden` attribute is added.

New example: https://live.bootstrap-table.com/code/marceloverdijk/3426 (using local data from the html document)

But also the initial [example](https://live.bootstrap-table.com/code/marceloverdijk/3261) is now working /cc @UtechtDustin 


Note while working on it I noticed:

https://github.com/wenzhixin/bootstrap-table/blob/develop/src/extensions/filter-control/bootstrap-table-filter-control.js#L484-L486

`UtilsFilterControl.getControlContainer(that).find('.filterControl').hide()` 
shouldn't this be:
`UtilsFilterControl.getControlContainer(that).find('.filter-control').hide()`?

File: site/docs/extensions/filter-control.md [hunk 1]
```diff
@@ -33,6 +33,19 @@ Dependence if you use the datepicker option: [bootstrap-datepicker](https://gith
 
 - **Default:** `false`
 
+### filterControlHidden
```
Comment (wenzhixin) [inline site/docs/extensions/filter-control.md:36]: I think `filterControlVisible` is better than `filterControlHidden`,  like our ohter options.

File: src/extensions/filter-control/bootstrap-table-filter-control.js [hunk 2]
```diff
@@ -958,4 +964,15 @@ $.BootstrapTable = class extends $.BootstrapTable {
       }
     }
   }
+
+  toggleFilterControl () {
+    this.options.filterControlHidden = !this.options.filterControlHidden
+    const filterControl = UtilsFilterControl.getControlContainer(this).find('.filter-control')
```
Comment (wenzhixin) [inline src/extensions/filter-control/bootstrap-table-filter-control.js:970]: `const $filterControls`

**Comment by UtechtDustin** [comment]
Please also (re)set the correct table header height.
https://github.com/wenzhixin/bootstrap-table/blob/develop/src/extensions/filter-control/bootstrap-table-filter-control.js#L84

File: site/docs/extensions/filter-control.md (hunk 1 above)
Comment (marceloverdijk) [reply site/docs/extensions/filter-control.md:36]: OK, will change it later today.

**Comment by marceloverdijk** [comment]
> Please also (re)set the correct table header height.
> https://github.com/wenzhixin/bootstrap-table/blob/develop/src/extensions/filter-control/bootstrap-table-filter-control.js#L84

@UtechtDustin where should that be called? Currently I see only here: https://github.com/wenzhixin/bootstrap-table/blob/develop/src/extensions/filter-control/bootstrap-table-filter-control.js#L668

**Comment by marceloverdijk** [comment]
> > Please also (re)set the correct table header height.
> > https://github.com/wenzhixin/bootstrap-table/blob/develop/src/extensions/filter-control/bootstrap-table-filter-control.js#L84
> 
> @UtechtDustin where should that be called? Currently I see only here: https://github.com/wenzhixin/bootstrap-table/blob/develop/src/extensions/filter-control/bootstrap-table-filter-control.js#L668

Never mind I see it now in the example. Will add it later today.

**Comment by marceloverdijk** [comment]
`filterControlHidden` is now renamed to `filterControlVisible` (with default value `true`).
But I'm struggling with that height fix and css...

**Comment by UtechtDustin** [comment]
@marceloverdijk 
Adjust the `fixHeaderCSS` function, check if filter-control is visible or not.
If it is visble: use the current code, if it is not visible: unset the height.

Now you can call the fixHeaderCSS function while toggling.

**Comment by marceloverdijk** [comment]
> @marceloverdijk
> Adjust the `fixHeaderCSS` function, check if filter-control is visible or not.
> If it is visble: use the current code, if it is not visible: unset the height.
> 
> Now you can call the fixHeaderCSS function while toggling.

I think the `fixHeaderCSS` is only called if the table `height` is set, right?
But filter control with a fixed height seems to be broken... https://live.bootstrap-table.com/code/marceloverdijk/3439 (using `1.16.0`)


And we also have mis-alignment @djhvscf [mentioned](https://github.com/wenzhixin/bootstrap-table/pull/5069#pullrequestreview-430741014).

**Comment by UtechtDustin** [comment]
@marceloverdijk 
Oh okay, i dont checked yet the function (i thought its called always).
Then you have to find a other way, maybe some of us three has time to help out the next days.

Yes the table is broken with a fixed height #4869

**Comment by marceloverdijk** [comment]
@UtechtDustin think I found it :-)

For each column that has no filter control an empty div is created with `.no-filter-control`. As this takes up some space we got the misalignment.

On creation and toggle I'm now not only hiding/showing the `.filter-control` elements, but also the `.no-filter-control` elements and now the misalignment is gone.

Note there is a bit of misalignment, but I see that also with `1.16.0` ([example](https://live.bootstrap-table.com/code/marceloverdijk/3443)).

Here are also updated example links:

- https://live.bootstrap-table.com/code/marceloverdijk/3261 (default)
- https://live.bootstrap-table.com/code/marceloverdijk/3442 (visible = false)

Thanks for your patience btw 👍 




PS: the more I looked at the code I wonder if this is correct:
https://github.com/wenzhixin/bootstrap-table/blob/develop/src/extensions/filter-control/bootstrap-table-filter-control.js#L484-L486
The class name seems incorrect, and should't this hide the `.no-filter-control` elements as well?

**Comment by UtechtDustin** [comment]
@marceloverdijk Thank you for your contributen and the good communication :-)
Yes there is the wrong class on L485, you can fix that.

After this is fixed i check your code/changes :-)

**Comment by marceloverdijk** [comment]
Cool I changed L485 from `UtilsFilterControl.getControlContainer(that).find('.filterControl').hide()` to `UtilsFilterControl.getControlContainer(that).find('.filter-control, .no-filter-control').hide()`

**Comment by UtechtDustin** [comment]
Before: https://live.bootstrap-table.com/code/UtechtDustin/3444
After: https://live.bootstrap-table.com/code/UtechtDustin/3445

File: site/docs/extensions/filter-control.md [hunk 3]
```diff
@@ -29,10 +29,22 @@ Dependence if you use the datepicker option: [bootstrap-datepicker](https://gith
...
 
+### filterControlVisible
+
+- **Attribute:** `data-filter-control-visible`
+
+- **type:** `Boolean`
+
+- **Detail:**
+
+   Set to false to hide the filter controls.
```
Comment (UtechtDustin) [inline site/docs/extensions/filter-control.md:44]: Set to `false` to hide the filter controls.

File: site/docs/extensions/filter-control.md [hunk 4]
```diff
@@ -254,6 +266,10 @@ Dependence if you use the datepicker option: [bootstrap-datepicker](https://gith
 
 * Clear all the controls added by this plugin (similar to showSearchClearButton option).
 
+### toggleFilterControl
+
+* Toggle filter control.
```
Comment (UtechtDustin) [inline site/docs/extensions/filter-control.md:271]: Some more info would be greate e.g.   
Toggles the filter controls (Toggles the visibility of the filter controls)

**Comment by marceloverdijk** [comment]
@UtechtDustin Documentation updated as well. 
After this is merge I will look into #5065.

**Comment by djhvscf** [comment]
Good work! Thanks

**Comment by marceloverdijk** [comment]
Thanks a lot for merging this! I will look into #5065 after work this evening.

**Pull request merged**